public/app/features/explore/state/history.ts

```diff
@@ -140,6 +145,20 @@ export const clearRichHistoryResults = (exploreId: ExploreId): ThunkResult<void>
  */
 export const initRichHistory = (): ThunkResult<void> => {
   return async (dispatch, getState) => {
+    const queriesMigrated = store.getBool(RICH_HISTORY_SETTING_KEYS.migrated, false);
+    const migrationFailedDuringThisSession = getState().explore.richHistoryMigrationFailed;
+
+    // Query history migration should always be successful, but in case of unexpected errors we ensure
+    // the migration attempt happens only once per session, and the user is informed about the failure
+    // in a way that can help with potential investigation.
+    if (config.queryHistoryEnabled && !queriesMigrated && !migrationFailedDuringThisSession) {
```

Comment:
Just wondering if we might want to also check that there was something in local storage to migrate first. Wondering about users who don't keep anything in local storage, can only use incognito tabs or whatever else. It wouldn't be that much overhead to start a migration that didn't have anything, but I think we should be able to check a variable quick here before proceeding

Reply:
Good point. In incognito mode it'd make redundant calls to migrate. Updated in e7b5aa3ae8e5b7999a27eccc3d909d56613039bf